ipi/engine/forcefields.py

```diff
@@ -446,6 +447,114 @@ def evaluate(self, r):
+    ):
+        """Initialises FFdmd.
+
+        Args:
+           pars: Optional dictionary, giving the parameters needed by the driver.
+        """
+
+        # check input - PBCs are not implemented here
+        #        if dopbc:
+        #            raise ValueError("Periodic boundary conditions are not supported by FFdmd.")
+
+        # a socket to the communication library is created or linked
+        super(FFdmd, self).__init__(latency, name, pars, dopbc=dopbc, threaded=threaded)
+
+        if coupling is None:
+            raise ValueError("Must provide the couplings for DMD.")
+        if freq is None:
+            raise ValueError(
+                "Must provide a frequency for the periodically oscillating potential."
+            )
+        if dtdmd is None:
+            raise ValueError(
+                "Must provide a time step for the periodically oscillating potential."
+            )
+        self.coupling = coupling
+        self.freq = freq
+        self.dtdmd = dtdmd
+        self.dmdstep = dmdstep
+
+    def poll(self):
+        """Polls the forcefield checking if there are requests that should
+        be answered, and if necessary evaluates the associated forces and energy."""
+
+        # We have to be thread-safe, as in multi-system mode this might get
+        # called by many threads at once.
+        with self._threadlock:
+            for r in self.requests:
+                if r["status"] == "Queued":
+                    r["status"] = "Running"
+                    r["t_dispatched"] = time.time()
+                    self.evaluate(r)  # MR BAD
+
+    def evaluate(self, r):
+        """Evaluating dmd: pbc: YES,
+        cutoff: NO, neighbour list: NO."""
+
+        q = r["pos"].reshape((-1, 3))
+        nat = len(q)
+        cell_h, cell_ih = r["cell"]
+
+        if len(self.coupling) != int(nat * (nat - 1) / 2):
+            raise ValueError("Coupling matrix size mismatch")
+
+        v = 0.0
+        f = np.zeros(q.shape)
+        # must think and check handling of time step
+        periodic = np.sin(self.dmdstep * self.freq * self.dtdmd)
+        # MR: the algorithm below has been benchmarked against explicit loop implementation
+        for i in range(1, nat):
+            # MR's first implementation:
+            #            dij = q[i] - q[:i]
+            #            rij = np.sqrt((dij ** 2).sum(axis=1))
+            # KF's implementation:
+            dij, rij = vector_separation(cell_h, cell_ih, q[i], q[:i])
+            cij = self.coupling[i * (i - 1) // 2 : i * (i + 1) // 2]  #
+            prefac = np.dot(
+                cij, rij
+            )  # for each i it has the distances to all indexes previous
+            v += np.sum(prefac) * periodic
+            dij *= -(cij / rij)[:, np.newaxis]  # magic line...
+            f[i] += dij.sum(axis=0) * periodic
+            f[:i] -= dij * periodic  # everything symmetric
+
+        r["result"] = [v, f.reshape(nat * 3), np.zeros((3, 3), float), ""]
```

Comment:
Can you please also add a virial calculator, or have a way of raising an error ?

Reply:
I'll  add the virial, it seems very easy for this specific ff. Thanks. However, should I also add a warning then for other current ff's that assign only 0 to the virials? (e.g. Debye and LJ)?

Reply:
Not yet ready but on the way.

Reply:
ok